File: week6/pset6/credit/credit.py
from cs50 import get_int
import math


def main():

    # Get credit card integer from user, convert to string
    creditCard = str(get_int("Credit card number: "))

    length = len(creditCard)

    finalSum = 0

    checkSum = False

    # Loop through the credit card string, start at the 2nd last digit
    for i in range(length - 2, -1, -2):

        # Multiply each of those digits by 2
        i = int(creditCard[i]) * 2

        # If, after multiplication, that number is two digits long, split the digits and then add them together
        if i > 9:
            a = math.trunc(i / 10)
            b = i % 10
            i = a + b

        # Add the resulting number to the finalSum variable
        finalSum += i

    # Go through every left over number in the credit card string, and add them to the finalSum variable
    for i in range(length - 1, -1, -2):

        finalSum += int(creditCard[i])

    # Check if the final digit of the finalSum variable is zero, if true set checkSum to true
    if finalSum % 10 == 0:

        checkSum = True

    # Print out the type of card
    if checkSum:

        if creditCard[0] == '4':
            print("VISA")
        elif creditCard[0] == '5':
            print("MASTERCARD")
        elif creditCard[0] == '3':
            print("AMEX")

    else:
        print("INVALID")


main()

File: week1/pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>
#include <string.h>

int main(void)
{

    long int credit = get_long("Number: ");

    int firstSum;
    int finalSum;

    bool checkSum = false;
    bool cardChecked = false;

    // Takes every second number, starting with the second from last, multiplies them by 2, and returns the sum.
    for (long int i = 10; i < credit; i *= 100)
    {

        int firstSumNumber = (trunc((credit / i) % 10)) * 2;

        // If any of the sums are a two digit number, split them into individual digits and add them.
        if (firstSumNumber > 9)
        {
            int a = trunc(firstSumNumber / 10);
            int b = firstSumNumber % 10;
            firstSumNumber = (a + b);
        }

        firstSum = firstSum + firstSumNumber;

    }

    finalSum = firstSum;

    // Takes all remaining numbers from the credit card and finds their sum.
    for (long int j = 1; j < credit; j *= 100)
    {

        finalSum += (trunc((credit / j) % 10));

    }

    // If the final digit of this last sum is equal to 0, the credit card is valid.
    if (finalSum % 10 == 0)
    {
        checkSum = true;
    }

    if (checkSum == true)
    {
        // AMEX check
        // Check if the first digit is 3, the second digit is 4 or 7, and if the length of the card number is 15 digits
        if (trunc(credit / pow(10, 14)) == 3)
        {
            int amexCheck = trunc(credit / pow(10, 13));

            if (amexCheck % 10 == 4 || amexCheck % 10 == 7)
            {
                printf("AMEX\n");
                cardChecked = true;
            }
            else
            {
                printf("INVALID\n");
                cardChecked = true;
            }
        }

        // Mastercard check
        // Check whether the first digit is 5, the second digit is between 1 and 5, and if the length of the card number is 16 digits
        if (trunc(credit / pow(10, 15)) == 5)
        {
            int mastercardCheck = trunc(credit / pow(10, 14));

            if (mastercardCheck % 10 > 0 && mastercardCheck % 10 < 6)
            {
                printf("MASTERCARD\n");
                cardChecked = true;

            }
            else
            {
                printf("INVALID\n");
                cardChecked = true;

            }
        }

        // Visa check
        // Check whether the first digit is 4, and if the length of the card number is 13 or 16 digits
        if (trunc(credit / pow(10, 12)) == 4 || trunc(credit / pow(10, 15)) == 4)
        {
            printf("VISA\n");
            cardChecked = true;
        }

        // Sets a back up value if the number passes the sumcheck test, but doesn't match any of the cards
        if (cardChecked == false)
        {
            printf("INVALID\n");
        }

    }
    else
    {
        printf("INVALID\n");
    }
}



